### PR TITLE
fix(v2): ignore import declarations in excerpt

### LINKED ISSUE #2373: '@docusaurus/useBaseUrl' showing in previous when sharing a link to docs on social media.
We must include:
`import useBaseUrl from '@docusaurus/useBaseUrl';`
...in order to use images in the pages.

When I share a docs page that includes this import, the preview (generated by Facebook, Whatsapp etc) shows this as the first line of content which looks poor.

image

I am using: 

`"dependencies": {
    "@docusaurus/core": "^2.0.0-alpha.43",
    "@docusaurus/preset-classic": "^2.0.0-alpha.43",
    "classnames": "^2.2.6",
    "react": "^16.13.0",
    "react-dom": "^16.13.0"
  },`

### LINKED ISSUE #2601: Bug (v2): Meta description outputting MDX import statements
## 🐛 Bug Report

MDX import statements which are required to be placed on the first line of a doc (i.e. after the doc header section), are being used in meta descriptions.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## To Reproduce

1. Create a new doc item.
1. Write the necessary header section.
1. Directly after the header section, add an import statement such as `import MyComponent from '../src/components/MyComponent';`.
1. Add some text below the import statement.
1. Visit the new doc item in your browser.
1. View the page source and see that the meta descriptions all use the import statement and not the 'actual' text you wrote in your doc file.

## Expected behavior

I expected to see the first line of my text in de meta descriptions.

## Actual Behavior

The import statement is being used in the meta descriptions.

## Your Environment

- Docusaurus version used: v2.0.0-alpha.50
- Environment name and version (e.g. Chrome 78.0.3904.108, Node.js 10.17.0): Firefox 74.0.1, Node.js v10.19.0
- Operating system and version (desktop or mobile): MacOSX High Sierra 10.13.6 (desktop)

## Reproducible Demo

(Paste the link to an example repo, including a `siteConfig.js`, and exact instructions to reproduce the issue.)

https://github.com/Yoast/developer-site combined with https://github.com/yoast/developer-docs/

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

Resolve #2373 
Resolve #2601

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Compare MDX page with `import`s, eg:

| Before   | After    |
| -------- | -------- |
| Page: https://deploy-preview-2380--docusaurus-2.netlify.com/docs/migrating-from-v1-to-v2 <br>  Description: import ColorGenerator from '@site/src/components/ColorGenerator'; | Page: https://v2.docusaurus.io/docs/migrating-from-v1-to-v2/ <br>  Description: This doc guides you through migrating an existing Docusaurus 1 site to Docusaurus 2.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus-utils/src/index.ts [hunk 1]
```diff
@@ -191,10 +193,15 @@ export function parse(
 } {
   const options: {} = {
     excerpt: (file: matter.GrayMatterFile<string>): void => {
-      file.excerpt = file.content
-        .trim()
-        .split('\n', 1)
-        .shift();
+      let fileContent = file.content.trimLeft();
+
+      if (RegExp(importRegexString).test(fileContent)) {
```
Comment (lex111) [inline packages/docusaurus-utils/src/index.ts]: For better performance, I first check to see if import declarations are at the very beginning of the file and only then remove them. Perhaps there is a better way (?)

Comment (yangshun) [reply packages/docusaurus-utils/src/index.ts]: Ideally we remove on the AST layer and not use regex, which can be unreliable. There can also be dynamic imports and CommonJS requires.

Comment (lex111) [reply packages/docusaurus-utils/src/index.ts]: Okay, should I close this PR?

Comment (lex111) [reply packages/docusaurus-utils/src/index.ts]: Honestly, we do not operate MDX AST in metadata, so I have no idea how to do it as you suggest. This case (remove `import`s) covers the main use case, and therefore resolves the issue..

Comment (lex111) [reply packages/docusaurus-utils/src/index.ts]: By the way, we still have a similar issue with MD rendering, for example https://v2.docusaurus.io/docs/lifecycle-apis

image

I have a solution for this, but it is in the same spirit as in current PR, so you are unlikely to like it.

**Comment by lex111** [comment]
But people _usually_ import components at the very beginning of the file, so we are only interested in this place. And given the current architecture, this is a not bad solution.
I also do not consider this patch something like hack, since we need to prepare a description from the MD markup, and accordingly we need to pre-process it. Just the hack is that users will have to move import declarations to another place in the file to get rid of _this bug_.

**Comment by lex111** [comment]
Should I close this PR?

**Comment by yangshun** [comment]
Ok let's close this. Sorry about this.

**Comment by lex111** [comment]
No problem, although I did not understand why this is a bad way, given that we need to remove the imports declarations at the top of the file, in the same way we could remove Markdown markup.

**Comment by lex111** [comment]
Reopen after discussion in #2601

cc @yangshun

**Pull request merged**